irohad/validation/impl/stateful_validator_impl.cpp

```diff
@@ -53,41 +73,54 @@ namespace iroha {
       // Filter only valid transactions
       auto filter = [&temporaryWsv, checking_transaction](auto &acc,
                                                           const auto &tx) {
-        auto answer = temporaryWsv.apply(tx, checking_transaction);
+        auto answer =
+            temporaryWsv.apply(*tx->makeOldModel(), checking_transaction);
         if (answer) {
           acc.push_back(tx);
         }
         return acc;
       };
 
-      auto &txs = proposal.transactions;
+      auto &txs = proposal->transactions();
       decltype(txs) valid = {};
 
-      model::Proposal validated_proposal(
-          std::accumulate(txs.begin(), txs.end(), valid, filter));
-      validated_proposal.height = proposal.height;
+      auto valid_txs = std::accumulate(txs.begin(), txs.end(), valid, filter);
+
+      // rework validation logic, so that this conversion is not needed
+      auto valid_proto_txs =
+          valid_txs
+          | boost::adaptors::transformed([](const auto &polymorphic_tx) {
+              return *static_cast<const shared_model::proto::Transaction *>(
+                  polymorphic_tx.operator->());
+            });
+      auto validated_proposal = proto::ProposalBuilder()
+                                    .createdTime(proposal->created_time())
+                                    .height(proposal->height())
+                                    .transactions(valid_proto_txs)
+                                    .build();
+
       log_->info("transactions in verified proposal: {}",
-                 validated_proposal.transactions.size());
-      return validated_proposal;
+                 validated_proposal.transactions().size());
+      return shared_model::detail::makePolymorphic<decltype(
```

Comment:
Use `interface::Proposal` instead, it isn't deduced from args

Reply:
`makePolymorphic` also does `make_shared`, which will not work on abstract type.

Reply:
Yeah, sorry. I mean proto::Proposal, ofc